README.md
# autohost
Convention-based, opinionated HTTP server library based on express. Lovingly ripped from the guts of Anvil.

## Rationale
As more services are introduced to a system, the tedium of fitting together all the same libraries over and over:

 * is soul-draining
 * encourages copy/pasta
 * adds inertia across multiple projects
 * increases the surface area for defects and maintenance

I created autohost so we could have a consistent, reliable and extendible way to create HTTP/socket powered sites and services. By introducing conventions and structure to projects, route definitions and handlers aren't scattered throughout the source and mixed with application logic.

## Features

 * Resource-based: define transport-agnostic resources that interact via HTTP or WebSockets
 * Supports server-side websockets and socket.io
 * Supports multiple Passport strategies via a pluggable auth provider approach

> __Note__

> The dashboard and related APIs are no longer included with autohost. They have been moved to a separate project: [autohost-admin](https://github.com/LeanKit-Labs/autohost-admin).

## Quick Start

	npm init
	npm install autohost autohost-nedb-auth -S

#### ./index.js - the most boring app ever
```js
var autohost = require( 'autohost' );
var auth = require( 'autohost-nedb-auth' )( {} );
var host = autohost( { authProvider: auth } );
// additional setup, like custom middleware would go here
host.start(); // starts the server
```

	node index.js

## autohost( config )
Refer to the section below for a list of available configuration properties and default values.

### Configuration
The object literal follows the format:

```js
// default shown for each property
{
	static: './public', 	// where to host static resources from
	anonymous: [], 			// add paths or url patterns that bypass authentication and authorization,

	port: 8800, 			// host port
	urlPrefix: undefined, 	// applies a global prefix to all routes - for use behind reverse proxy
	apiPrefix: '/api', 		// changes the prefix for resource action URLs only

	resources: './resource', // where to load resource modules from
	modules: [], 			// list of npm resource modules to load

	authProvider: undefined, // a promise for or instance of an authentication provider
	allowedOrigin: , 		// used to filter incoming web socket connections based on origin
	socketIO: false, 		// enables socket.io - can be false, true (using default options) or an options object
	websocket: false, 		// enables websockets

	noBody: false, 			// disables body parsing
	noCookie: false, 		// disables cookies
	noCrossOrigin: false, 	// disables cross origin
	noOptions: false, 		// disables automatic options middleware
	noProxy: false, 		// disables trusted proxies
	noSession: false, 		// disables sessions

	session: 				// session configuration
	cookie: 				// session cookie configuration
	getUserString:			// method to return a string id for a user

	enableAccessLogs: true,// enables access logs
	logging: {}, 			// configuration passed to autohost's whistlepunk instance
	fount: undefined, 		// pass the app's fount instance to autohost

	parseAhead: false, 			// parses path parameters before application middleware
	handleRouteErrors: false, 	// wrap routes in try/catch
	urlStrategy: undefined 		// a function that generates the URL per resource action
}
```

#### Socket.io
The configuration options (above) value for `socketIO` can be `false` (disabled), `true` (enabled with basic socket.io defaults) or an `object` specifying any [option supported by socket.io](http://socket.io/docs/server-api/#server(opts:object)).

#### Session Configuration
By default [express session](https://github.com/expressjs/session) is the session provider. To change any settings for how the session is configured, provide a hash with values for any of the properties shown below.

```javascript
// default shown for each property
{
	name: 'ah.sid',
	secret: 'autohostthing',
	resave: true,
	store: new sessionLib.MemoryStore(),
	saveUninitialized: true,
	rolling: false
}
```

This example demonstrates using the redis and connect-redis libraries to create a redis-backed session store.
```javascript
var autohost = require( 'autohost' );
var auth = require( 'autohost-nedb-auth' )( {} );

var redis = require( 'redis' ).createClient( port, address );
var RedisStore = require( 'connect-redis' )( host.session );
var store = new RedisStore( {
		client: redis,
		prefix: 'ah:'
	} );

host = autohost( {
	authProvider: auth,
	session: {
		name: 'myapp.sid',
		secret: 'youdontevenknow',
		store: store
	}
} );
host.start();
```

#### Ending a session
To end a session:

 * `logout` method on the envelope in a resource action handle
 * `logout` on the request in any middleware

#### Session Cookie Configuration
To change any settings for how the session **cookie** is configured, provide a hash with values for any of the properties shown below.

```javascript
// default shown for each property
{
	path: '/',
	secure: false,
	maxAge: null
}
```

#### Static

The static option supports either a path, an options hash, or `false`. Currently, the options (except `path`) are passed through to [`express.static`](http://expressjs.com/guide/using-middleware.html#express.static) with the `path` property being used as the route. (If set to `false`, no default static path will be auto-configured):

```js
{
	static: {
		path: './public',
		maxAge: '2d',
		setHeaders: function ( res, path, stat ) { ... }
	}
}
```

### AuthProvider
There are already two available auth provider libraries available:
 * [autohost-riak-auth](https://github.com/LeanKit-Labs/autohost-riak-auth)
 * [autohost-nedb-auth](https://github.com/LeanKit-Labs/autohost-nedb-auth)

Each library supports all optional features and can be managed from the [admin add-on](https://github.com/LeanKit-Labs/autohost-admin).

	Note: the authProvider passed in can be an unresolved promise, autohost will handle it

Planned support for:
 * MS SQL server

### getUserString

 The `getUserString` option expects a method that accepts `user` as its only parameter, and returns a string (used for logging) to identify the user. The default method provided attempts the following steps:

 * return `user.name` if available, otherwise:
 * return `user.username` if available, otherwise:
 * return `user.id` if available, otherwise:
 * return `JSON.stringify( user )`

 Override this method with custom logic if the default does not match your field names on your user object. For instance:

 ```js
 {
 	getUserString: function ( user ) {
 		return user.login;
 	}
 }
 ```

### fount
[fount](https://github.com/LeanKit-Labs/fount) is a dependency injection library for Node. If the application is using fount, the application's instance can be provided at the end of the init call so that resources will have access to the same fount instance the application is using. The fount instance in use by `autohost` is available via `host.fount`.

## Resources
Resources are expected to be simple modules containing a factory method that return one or more resource definitions. Dependency resolution by argument is supported in these resource factory methods. All arguments after the first (`host`) will be checked against autohost's fount instance. This is especially useful when taking a dependency on a promise or asynchronous function. Fount will only invoke the resource's factory once all dependnecies are available, eliminating dependency callbacks or promises in the resource's implementation. See the Asynchronous Module example under the Module section.

### Path conventions
All resources must be placed under a top level folder (`./resource` by default) and shared static resources under a top level folder (`./public` by default). Each resource should have its own sub-folder and contain a `resource.js` file that contains a module defining the resource.

####Folder structure
	-myProject
	 `-- resource
	 |	`-- profile
	 |  |  |-- resource.js
	 |  |
	 |  `-- otherThing
	 |  |  |-- resource.js
	 `--public
	 |  `--css
	 |  |  |--main.css
	 |  `--js
	 |  |  |--jquery.min.js
	 |  |  |--youmightnotneed.js
	 |  |--index.html

####Module
__Synchronous Module - No Fount Dependencies__
```js
module.exports = function( host ) {
	return {
		name: 'resource-name',
		static: '', // relative path to static assets for this resource,
		apiPrefix: '', // Optional override for global apiPrefix setting. Omit entirely to use default.
		urlPrefix: '', // URL prefix for all actions in this resource
		middleware: [],// one or more middleware functions to mount to the resource's url
		actions:  {
			send: {
				method: 'get', // http verb
				url: '', // url pattern appended to the resource name
				topic: 'send', // topic segment appended the resource name
				middleware: [], // one or more middleware functions to mount to the action's url
				authorize: , // optional predicate to check user permissions
				handle: function( envelope ) {
					// see section on envelope for more detail
				}
			}
		}
	};
};
```

__Asynchronous Module - Fount Dependencies__
This example assumes that either:

 * the application fount instance was plugged into autohost or
 * all defined dependencies were made with autohost's fount instance before calling autohost's `init` call.

```js
// example using autohost's fount instance
var autohost = require( 'autohost' );
var host = autohost( { ... } );
host.fount.register( 'myDependency1', { ... } );
host.fount.register( 'myDependency2', somePromise );

```

```js
// Each argument after `host` will be passed to fount for resolution before the exported function
// is called.
module.exports = function( host, myDependency1, myDependency2 ) {
	return {
		name: 'resource-name',
		static: '', // relative path to static assets for this resource
		apiPrefix: '', // Optional override for global apiPrefix setting. Omit entirely to use default.
		urlPrefix: '', // URL prefix for all actions in this resource
		middleware: [],// one or more middleware functions to mount to the resource's url
		actions: {
			send: {
				method: 'get', // http verb
				url: '', // url pattern appended to the resource name
				topic: 'send', // topic segment appended the resource name
				middleware: [], // one or more middleware functions to mount to the action's url
				authorize: , // optional predicate to check user permissions
				handle: function( envelope ) {
					// see section on envelope for more detail
				}
			}
		]
	};
};
```

### name
The resource name is pre-pended to the action's alias to create a globally unique action name: `resource-name.action-alias`. The resource name is also the first part of the action's URL (after the api prefix) and the first part of a socket message's topic:

	http://{host}:{port}/api/{resource-name}/{action-alias|action-path}

	topic: {resource-name}.{action-topic|action-alias}

	Note: If defining resources for use with [hyped](https://github.com/leankit-labs/hyped) - the resource name is not automatically pre-pended to the url.

### static
You can host nested static files under a resource using this property. The directory and its contents found at the path will be hosted after the resource name in the URL.

### middleware
Provides a mechanism for defining resource-level middleware either in a single function or in a list of functions. These functions are provided with an envelope and are able to make changes to it before it reaches action-specific middleware or the handle call.

Below are several examples of different middleware patterns. This should demonstrate both synchronous and asynchronous patterns for proceding and short-circuiting the stack.

```javascript
...
// all middleware must return either the result of next or a promise/data structure
middleware: [
	function( envelope, next ) {
		// invokes the next middleware or handle call
		return next();
	},
	function( envelope, next ) {
		// demonstrates returning a data structure
		if( envelope.data.example === 1 ) {
			return { data: { message: 'This will short circuit the stack and respond immediately' } };
		} else if( envelope.data.example === 2 ) {
			// demonstrates returning next asynchronously
			return somethingPromisey()
				.then( function( x ) {
					envelope.context.importantThing = x;
					return next();
				} );
		} else if( envelope.data.example === 3 ) {
			// demonstrates short-circuiting a stack via a promise
			return anotherPromise()
				.then( function( x ) {
					return { data: x };
				} );
		} else {
			return next();
		}
	}
}
...
```

## Actions
The hash of actions are the operations exposed on a resource on the available transports.

### [key]
They key of the action in the hash acts as the 'friendly' name for the action. To create a globally unique action name, autohost pre-pends the resource name to the alias: `resource-name.action-alias`.

### method
Controls the HTTP method an action will be bound to.

### topic
This property controls what is appended to the resource name in order to create a socket topic. The topic is what a socket client would publish a message to in order to activate an action.

### url - string pattern
The `url` property provides the URL assigned to this action. You can put path variables in this following the express convention of a leading `:`

	url: '/thing/:arg1/:arg2'

Path variables are accessible on the envelope's `params` property. If a path variable does NOT collide with a property on the request body, the path variable is written to the `envelope.data` hash as well:

```javascript
	envelope.data.arg1 === envelope.params.arg1;
```

### url - regular expression
The `url` can also be defined as a regular expression that will be evaluated against incoming URLs. Both `apiPrefix` and `urlPrefix` will be pre-pended to the regular expression automatically - do not include them in the expression provided.

#### query parameters
Query parameters behave exactly like path variables. They are available on the `params` property of the envelope and copied to the `envelope.data` hash if they wouldn't collide with an existing property.

#### custom url strategy
A function can be provided during configuration that will determine the url assigned to an action. The function should take the form:

```javascript
function myStrategy( resourceName, actionName, action, resourceList ) { ... }
```

The string returned will be the URL used to route requests to this action. Proceed with extreme caution.

### middleware
Provides a mechanism for defining action-level middleware either in a single function or in a list of functions. These functions are provided with an envelope and are able to make changes to it before it reaches the handle call.

__IMPORTANT__
Middleware **must** return either the result of the `next` call _or_ a promise/data structure to short circuit the stack with a response. They are mutually exclusive. Do not call both. Do not fail to return one or the other.

### authorize
The `authorize` predicate was added to actions to allow for much more fine grained, explicit control of user authorization. By default, authorization checks are performed after _all_ middleware has run. This allows middleware to provide any necessary data on the envelope's context so that the authorization strategy can use this in determining the user's access level.

To change this, provide the string "authorize" in place of a middleware call in the resource or action middleware property and the check will be performed where the string appeared in the stack. This allows an application to short-circuit the stack and avoid running middleware that performs expensive and unnecessary i/o if the user does not have permission to perform the action.

> Note: when present, this _overrides_ rather than augments any authorization check that would have been performed by a configured autohost auth library.

```javascript
...
authorize: function( envelope ) {
	return envelope.user.isAwesome;
}
...
}
```

### handle
The handle is a callback that will be invoked if the caller has adequate permissions. The handle call can return a hash (or a promise that resolve to one) with the following properties:

> Note: `data`, `file`, `forward` and `redirect` are mutually exclusive. Websockets only supports `data` and `file`.

```javascript
// defaults shown
{
	status: 200,
	data: undefined,
	cookies: {}, // set cookies sent back to the client
	headers: {}, // set headers sent back in the response
	file: { // only used when replying with file
		name: , // the file name for the response
		type: , // the content-type
		stream: // a file stream to pipe to the response
	},
	forward: { // only used if forwarding the request
		url: , // the url to forward to
		method: , // if unspecified, copies the method of the original request
		headers: , // if unspecified, copies headers in the original request
		body: // use if changing the body contents
	},
	redirect: { // only used when redirecting
		status: 302, // use to set a status other than 302
		url: // the URL to redirect to
	}
}
```

> **Recommendation**

> Don't include application logic in a resource file. The resource is there as a means to 'plug' application logic into HTTP and websocket transports. Keeping behavior in a separate module will make it easy to test application behavior apart from autohost.

#### differentiated handlers
In rare cases, you may need the ability to have multiple handler behaviors that change based on some aspect of the request. Autohost allows a hash to specify multiple handlers and conditions for when each should be applied. The list is evaluated in order, keep this in mind when putting conditions in place. If no condition is met, a 400 will be returned - the assumption being that you, the developer intentionally excluded some condition and the caller submitted a malformed request.

> Note: this feature exists primarily so our upstream hypermedia library, `hyped` can plug in handlers based on version.

```javascript
{
	handle: [
		{
			when: { version: 1 }, // providing a set of properties and values to filter requests for the handler
			then: function( envelope ) {
				...
			}
		},
		{
			when: function( envelope ) { // provide a predicate to test the envelope
				return envelope.version === 2;
			},
			then: function( envelope ) {
				...
			}
		},
		{
			when: true, // use at the end as a catch-all if desired
			then: function( envelope ) {
				...
			}
		}
	]
}
```

### Controlling Error Responses
Responses sent to the client based on an error returned from an action's handle can be controlled at the config, resource or action level. How to handle a specific error type is determined by first checking the action, then resource, then config (host) levels.

The `errors` property can be set at any of these levels and is a set of case-sensitive error names and a literal specifying how to render the error. The literal can contain a `status` to control the status code used and a static `body`, `file` or `reply` function that takes the error as an argument and returns the content for the response body.

> Note: File is only applicable for the http transport and will be ignored in sockets.

```javascript
// this could exist in the config, a resource or an action
errors: {
	Error: {
		status: 500,
		body: 'oops'
	},
	NotFoundError: {
		status: 404,
		file: './404.html' // file is relative to the static folder
	},
	BadRequestError: {
		status: 400,
		reply: function( err ) {
			return 'This is no good: ' + err.message;
		}
	}
},
```

#### Tighter Response Control
Read the section on envelopes for details on data available and alternate ways to produce a response.

## Envelope
Envelopes are an abstraction around the incoming message or request. They are intended to help normalize interactions with a client despite the transport being used.

```javascript
// common properties/methods
{
	context: // metadata added by middleware
	cookies: // cookies on the request
	data: // the request/message body
	headers: // request or message headers
	logout: // a method to end the current session
	path: // url of the request (minus protocol/domain/port) OR message topic
	session: // session hash
	responseStream: // a write stream for streaming a response back to the client
	transport: // 'http' or 'websocket'
	version: // will check `req.context.version` - provide middleware to set this

CHANGELOG.md
## 3.x

### 3.0.1

* Removed node-uuid in favor of uuid

### 3.0.0

* Removed unused dependencies
* Updated multer
* Updated whistlepunk
* Updated passport and conditionally add passport session middleware based on changes to passport
* Updated socket.io
* Updated passport to 0.5.3
* Updated query-string dep
* Updated sinon/chai related deps
* Updated more dev deps
* Updated some dev deps
* Added a few updated dependencies
* Updated to latest semver versions
* Removed a couple of unneeded dependencies
* Updated node version

## 2.x

### 2.0.0

* Updated project dependencies and security issues

## 1.x

### 1.2.0

* Fixed tests for new enable value.
* Changed to an enable flag, enableAccessLogs, that defaults to true when undefined.
* Add option to disable access logs

### 1.1.0

* Named the main route handler function for better reporting in New Relic and similar tools

### 1.0.0

* change order of operations of merge operations into `data` to prefer url parameter, then request body, then parameter
* request body is deep cloned to the data propert on HttpEnvelope to avoid
* add query, parameter, and body properties to envelope

## 0.6.x

### 0.6.2
 * Bug fix - change order of merge operations into `data` to prefer url parameter, then request body, then query parameter.
 * Bug fix - clone request body to data property on envelope; do not copy reference and allow mutation
 * Improvement - add query, parameter and body properties to envelope

### 0.6.1

 * Add ability to define multiple URL patterns per action
 * Add support for differentiated authorization (support versioned authorization in hyped)

### 0.6.0

* Removed Express.js warning when not using sessions

## 0.5.x

### 0.5.12

* Updated socket.io to 1.4.5 and websocket to 1.0.22

### 0.5.11

 * Fixes around socket connection bookkeeping

### 0.5.10

 * Regex fix for newer version of node

### 0.5.9
 * Better default error logging for 5xx errors
 * Minor tweak to prevent posting two timing metrics for resource actions called via HTTP

### 0.5.8
 * Add differentiated handle feature (to support adding versioned handlers in hyped)

### 0.5.7

 * Fixed issue with data objects having no hasOwnProperty function

### 0.5.6
 * Provide a means to customize socket.io configuration
 * Update dependencies
   * change dependency on qs to query-string
   * update multer usage to adapt to new API

### 0.5.5
 * Allow auth errors to bubble up to the default error handler (or custom error handlers)

### 0.5.4
 * Fixed an issue where resource modules that return a promise for a resource were not having their relative path set

### 0.5.3
 * Fixed issue where cors config was not properly accessed in middleware

### 0.5.2
 * Added ability to configure Access-Allow headers via top level config
 * Added JSCS configuration file
 * Skipped auth on options calls

### 0.5.1
 * Bug fix - error results not getting rendered correctly from hyped
 * Change default error strategy to use an object with a `message` property to match other AH default responses
 * No longer return errors as message body (too dangerous a default behavior)
 * Add `method` property to httpEnvelope
 * Remove demo and dependencies on auth provider libraries

### 0.5.0
 * A query string value can no longer override a URL parameter (Matches behavior of the `data` object)
 * Correct issue where insufficient parameters were passed to authorize predicates in hyped
 * Default authorization (permission) checks to the end of all middleware
 * Allow a placeholder to control when authorization checks occur in middleware stacks
 * Return JSON `{ message: ... }` for all error responses
 * Fix issue where express middleware that responds with an error code did not end HTTP stack evaluation

### 0.4.7
 * Add authorize predicate to resource actions to support alternate authorization approach

### 0.4.6
 * Add resource and action middleware abstraction
 * Minor corrections to existing README for previous features

### 0.4.5
Enhancement - getUserString is now shared as an overrideable global config option

### 0.4.4
Bug fix - getRoles was incorrectly being treated as error handling middleware

### 0.4.3

 * Add metadata to metrics to provide a unique name for each metric collected
 * Bug fix - prevent default metronic instance from being used when user provides one via config

### 0.4.2

 * Bug fix - authenticated users don't show up in access log correctly on consecutive calls
 * Remove authenticatedUser property from request

### 0.4.1

 * API Change - host now requires `start` to start the server
 * authProvider is now a config property
 * Bug fix - undefined config would throw an error instead of using all defaults
 * Bug fix - correct registration of system middleware as user middleware
 * Remove support for incorrectly nested user property on request object
 * Added support for handling generic errors
 * Support returning static files as error responses
 * Added aliases for anonymous middleware for better debug logging
 * Rework harness to rely on index
 * Expose harness via 'autohost/harness'

### 0.4.0

 * Improve control over session and session cookie configuration
 * Bug fix - socket.io adapter re-attached authentication middleware on every connection
 * Minor improvement to logging abstraction and updates to specs still using debug
 * Refactored module-level state out of modules
 * Support object literal syntax for resource action handles
 * Support custom error handling

## 0.3.x
Primary motivation here is to begin work on a version of autohost that will work well with a hypermedia library ( [hyped](https://github.com/leankit-labs/hyped) ). This is a breaking change because of several structural and naming changes to how resources get modeled.

### 0.3.2

 * Bug fix - multiple modules were incorrectly passing config to metronic

### 0.3.1
 * Update metronic version to 0.2.0
 * Bug fix - resolving external resource modules should correctly differentiate between relative paths and NPM modules


src/autohost.js
// Forked from Anvil's http host and other modules
// Copyright Alex Robson 2011
// MIT License - http://opensource.org/licenses/MIT

var http = require( 'http' ),
	ServerResponse = require( 'http' ).ServerResponse,
	express = require( 'express' ),
	path = require( 'path' ),
	fs = require( 'fs' ),
	url = require( 'url' ),
	net = require ( 'net' ),
	_ = require( 'lodash' ),
	mkdirp = require( 'mkdirp' ),
	request = require( 'request' ),
	Monologue = require( 'monologue.js' )( _ ),
	gaze = require( 'gaze' ),
	SocketStream = require( './socketStream' ),
	metrics = require( 'cluster-metrics' );

module.exports = function( config ) {
	var PREFIX = 'autohost';

	var Host = function() {
		this.clients = [];
		this.clients.lookup = {};
		this.compilers = {};
		this.middleware = [];
		this.config = _.defaults( config, { apiPrefix: '/api' } );
		this.topics = {};
		this.resources = {};
		this.actions = {};
		this.metrics = metrics;
		this.appName = this.config.appName || 'autohost';
		_.bindAll( this );
		this.request = request.defaults( { jar: true } );
	};

	Host.prototype.addCompiler = function( extension, callback ) {
		this.compilers[ extension ] = callback;
		if( this.app ) {
			this.app.engine( extension, callback );
		}
	};

	Host.prototype.addMiddleware = function( filter, middleware ) {
		this.middleware.push( { filter: filter, handler: middleware } );
		if( this.app ) {
			this.app.use( filter, middleware );
		}
	};

	Host.prototype.buildPath = function( pathSpec ) {
		var hasLocalPrefix;
		pathSpec = pathSpec || '';
		if( _.isArray( pathSpec ) ) {
			hasLocalPrefix = pathSpec[0].match( /^[.]\// );
			pathSpec = path.join.apply( {}, pathSpec );
		}
		pathSpec = pathSpec.replace( /^~/, process.env.HOME );
		return hasLocalPrefix ? './' + pathSpec : pathSpec;
	};

	// yay, Jim Cowart to the rescue with this bit
	Host.prototype.buildUrl = function() {
		var idx = 0,
			cleaned = [];
		while( idx < arguments.length ) {
			var segment = arguments[ idx ];
			if( segment.substr( 0, 1 ) === '/' ) {
				segment = segment.substr( 1 );
			}
			if( segment.substr( segment.length-1, 1 ) === '/' ) {
				segment = segment.substring( 0, segment.length - 1 );
			}
			if( !_.isEmpty( segment ) ) {
				cleaned.push( segment );
			}
			idx++;
		}
		return '/' + cleaned.join( '/' );
	};

	Host.prototype.handleRoute = function( url, verb, resource, handle, action ) {
		this.registerRoute( url, verb, function( req, res ) {
			var respond = function() {
				var envelope = {
					context: req.context,
					data: req.body || {},
					path: req.url,
					cookies: req.cookies,
					headers: req.headers,
					params: {},
					files: req.files,
					user: req.user,
					responseStream: res,
					forwardTo: function( options ) {
						return req.pipe( this.request( options ) );
					}.bind( this ),
					reply: function( envelope ) {
						var code = envelope.statusCode || 200;
						res.send( code, envelope.data );
					},
					replyWithFile: function( contentType, fileName, fileStream ) {
						res.set( {
								'Content-Disposition': 'attachment; filename="' + fileName + '"',
								'Content-Type': contentType
							} );
						fileStream.pipe(res);
					}
				};
				for( var key in req.params ) {
					var val = req.params[ key ];
					if( !envelope.data[ key ] ) {
						envelope.data[ key ] = val;
					}
					envelope.params[ key ] = val;
				}
				for( var key in req.query ) {
					var val = req.query[ key ];
					if( !envelope.data[ key ] ) {
						envelope.data[ key ] = val;
					}
					envelope.params[ key ] = val;
				}
				handle.apply( resource, [ envelope ] );
			}.bind( this );

			if( this.authorizer ) {
				this.authorizer.getRolesFor( action, function( err, roles ) {
					roles = err ? [] : roles;

					if( roles.length > 0 && _.intersection( roles, req.user.roles ) == 0 ) {
						res.send( 403, "User lacks sufficient permissions" );
					} else {
						respond();
					}
				} );
			} else {
				respond();
			}
		}.bind( this ) );
	};

	Host.prototype.handleTopic = function( topic, resource, handle, action ) {
		this.registerTopic( topic, function( message, socket ) {
			var respond = function() {
				var envelope = {
					data: message.data || message,
					headers: message.headers || [],
					socket: socket,
					cookies: socket.cookies,
					user: socket.user,
					path: message.topic,
					reply: function( envelope ) {
						socket.publish( message.replyTo, envelope );
					},
					responseStream: new SocketStream( message.replyTo, socket )
				};
				handle.apply( resource, [ envelope ] );
			};
		
			if( this.authorizer ) {
				this.authorizer.getRolesFor( action, function( err, roles ) {
					roles = err ? [] : roles;
					if( roles.length > 0 && _.intersection( roles, socket.user.roles ) == 0 ) {
						socket.publish( 'NoPe', 'User lacks sufficient permissions' );
					} else {
						respond();
					}
				} );	
			} else {
				respond();
			}
		}.bind( this ) );
	};

	Host.prototype.init = function( config, done ) {
		if( config ) {
			this.config = _.merge( this.config, config );
		}
		var self = this,
			cwd = process.cwd(),
			tmp = path.resolve( cwd, ( this.config.temp || './tmp' ) ),
			public = path.resolve( cwd, ( this.config.static || './public' ) ),
			resources = path.resolve( cwd, ( this.config.resources || './resource' ) );

		//mkdirp( tmp );

		this.app = express();
		this.app.use( '/', function( req, res, next ) {
			req.context = {};
			var timerKey = [ req.method.toUpperCase(), req.url, 'timer' ].join( ' ' );
			metrics.timer( timerKey ).start();
			res.on( 'finish', function() { 
				metrics.timer( timerKey ).record();
			} );
			next();
		} );
		this.app.use( '/api', function( req, res, next ) {
			if( req.method == 'OPTIONS' || req.method == 'options' ) {
				this.resources.prefix = this.config.apiPrefix;
				res.send( 200, this.resources );
			} else {
				next();
			}
		}.bind( this ) );

		if( !config.noCookies ) {
			this.app.use( express.cookieParser() );
		}

		if( !config.noBody ) {
			this.app.use( express.bodyParser( {
				uploadDir: tmp,
				keepExtensions: true
			} ) );
		}

		if( !config.noSession ) {
			this.app.use( express.session( { secret: config.sessionSecret || 'authostthing' } ) );
		}

		if( !config.noCrossOrigin ) {
			this.app.use( function( req, res, next ) {
				res.header( 'Access-Control-Allow-Origin', '*' );
				res.header( 'Access-Control-Allow-Headers', 'X-Requested-With' );
				next();
			} );
		}
		
		this.addAuthentication();
		
		_.each( this.middleware, function( mw ) {
			self.app.use( mw.filter, mw.handler );
		} );

		var startServer = function() {
			this.server = http.createServer( this.app ).listen( this.config.port || 8800 );
			this.configureWebsockets();
			this.configureSocketIO();
			console.log( 'autohost listening on port ', ( this.config.port || 8800 ) );
			if (done) done();
		}.bind(this);

		this.app.use( this.app.router );
		this.registerPath( '/', public );
		this.loadResources( resources ).done(function () {
			this.processResource( this.config.apiPrefix, require( './_autohost/resource.js' )( this ), path.resolve( __dirname, './_autohost' ) );
			if( this.authorizer ) {
				var list = [];
				_.each( this.actions, function( actions, resource ) {
					_.each( actions, function( action ) {
						list.push( { name: action, resource: resource } );
					} );
				} );
				this.authorizer.actionList( list, function() {
					startServer();
				} );
			} else {
				startServer();
			}
		}.bind(this));
	};

	Host.prototype.registerRoute = function( url, verb, callback ) {
		verb = verb.toLowerCase();
		verb = verb == 'all' || verb == 'any' ? 'all' : verb;
		var routes = this.app.routes[ verb ],
			errors = [ PREFIX, url, verb, 'errors' ].join( '.' );
		if( routes ) {
			var index = routes.indexOf( url );
			if( index >= 0 )
			{
				console.log( 'remove duplicate route for ', url, verb );
				this.app.routes[ verb ].splice( index, 1 );
			}
		}
		if( this.app ) {
			this.app[ verb ]( url, function( req, res ) {
				try {
					callback( req, res );
				} catch ( err ) {
					metrics.meter( errors ).record();
					console.log( 'error on route, "' + url + '" verb "' + verb + '"', err.stack );
				}
			} );
		}
	};

	Host.prototype.registerPath = function( url, filePath ) {
		if( this.app ) {
			this.app.use( url, express[ 'static' ]( path.resolve( filePath ) ) );
		}
	};

	Host.prototype.registerTopic = function( topic, callback ) {
		var timer = [ PREFIX, topic, 'timer' ].join( '.' ),
			errors = [ PREFIX, topic, 'errors' ].join( '.' );
		this.topics[ topic ] = callback;
		_.each( this.clients, function( client ) {
			client.removeAllListeners( topic );
			client.on( topic, function( data ) {
				try {
					metrics.timer( timer ).start();
					callback( data, client );
					metrics.timer( timer ).record();
				} catch ( err ) {
					metrics.meter( errors ).record();
					console.log( 'error on topic, "' + topic + '"', err );
				}
			} );
		} );
	};

	Host.prototype.watch = function( filePath ) {
		var self = this;
		if( !fs.existsSync( filePath ) )
			return;
		return gaze( path.join( filePath, '**/resource.js' ), function( err, watcher ) {
			this.on( 'changed', function( changed ) {
				console.log( 'Reloading changed resource', path.basename( path.dirname( changed ) ) );
				self.loadModule( changed );
			} );
		} );
	};

	require( './passport.js' )( Host );
	require( './resource.js' )( Host );
	require( './websockets.js' )( Host );
	require( './socketio.js' )( Host );
	require( './resolver.js' )( Host );

	Monologue.mixin( Host );
	_.bindAll( Host );
	return new Host();
};